src/lib.rs

```diff
@@ -1,43 +1,400 @@
 //! # Geogram Predicates
 //!
-//! A crate for rust interoperability with `geogram`s _robust predicates_; via `cxx`.
+//! A Rust port of `geogram`s _robust predicates_.
+//! An interoperability ffi with `geogram`s _robust predicates_; via `cxx` also available in the `legacy` feature.
+//!
+//! All functions are in written in Rust, except for `det_3d` and `det_4d` which are only available in `legacy`.
+#![no_std]
+
+#![warn(clippy::all, unused, clippy::missing_const_for_fn)]
 
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "legacy")]
 pub use geogram_ffi::*;
 
-#[cxx::bridge(namespace = "GEOGRAM")]
-mod geogram_ffi {
-    // Shared structs with fields visible to both languages.
-    // ...
+use nalgebra::Point3;
+use robust::{Coord, Coord3D};
+
+mod types;
+pub use types::Sign;
+
+mod expansion;
+pub use expansion::expansion::Expansion;
+
+mod orient_2dlifted;
+pub use orient_2dlifted::orient_2dlifted_sos;
+
+mod orient_3dlifted;
+pub use orient_3dlifted::orient_3dlifted_sos;
+
+pub type Point3d = [f64; 3];
+pub type Point2d = [f64; 2];
+
+pub(crate) const FPG_UNCERTAIN_VALUE: Sign = Sign::Zero;
+
+/// Gets the sign of a value.
+///
+/// ### Parameters
+/// - `x` value to test
+///
+/// # Example
+/// ```
+/// use geogram_predicates::{Sign, geo_sign};
+///
+/// let a = 42.0;
+/// let b = -42.0;
+/// let c = 0.0;
+///
+/// assert_eq!(geo_sign(a), Sign::Positive);
+/// assert_eq!(geo_sign(b), Sign::Negative);
+/// assert_eq!(geo_sign(c), Sign::Zero);
+/// ```
+#[inline(always)]
+#[must_use]
+pub const fn geo_sign(value: f64) -> Sign {
+    if value > 0.0 {
+        Sign::Positive
+    } else if value < 0.0 {
+        Sign::Negative
+    } else {
+        Sign::Zero
+    }
+}
+
+/// Computes the orientation predicate in 3d.
+///
+/// Computes the sign of the signed volume of the tetrahedron `a`, `b`, `c`, `d`.
+///
+/// ### Parameters
+/// - `a`, `b`, `c`, `d` vertices of the tetrahedron
+///
+/// ### Return values
+/// * `Positive` - if the tetrahedron is oriented positively
+/// * `Zero` - if the tetrahedron is flat
+/// * `Negative` - if the tetrahedron is oriented negatively
+///
+/// # Example
+/// ```
+/// use geogram_predicates::orient_3d;
+///
+/// // Define four points that form a tetrahedron
+/// let a = [0.0, 0.0, 0.0];
+/// let b = [2.0, 0.0, 0.0];
+/// let c = [0.0, 2.0, 0.0];
+/// let d = [0.75, 0.75, 1.0];
+///
+/// assert_eq!(1, orient_3d(&a, &b, &c, &d));
+///```
+#[must_use]
+pub fn orient_3d(a: &Point3d, b: &Point3d, c: &Point3d, d: &Point3d) -> Sign {
+    let orientation = robust::orient3d(
```

Comment:
I see, we are using `robust` for all other predicates, which is okay for now.
But if you went ahead and implemented the `expansion` the heart of the algos, then we can now easily add all the other predicates, in a `geogram` fashion as well.
Also for a future PR!